File: 238-product-of-array-except-self/product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
    const n = nums.length;
    const ans = new Array(n).fill(0);

    let prefProd = 1;
    for (let i = 0; i < n; i++) {
        ans[i] = prefProd;
        prefProd *= nums[i];
    }

    let suffProd = 1;
    for (let i = n - 1; i >= 0; i--) {
        ans[i] *= suffProd;
        suffProd *= nums[i];
    }

    return ans;
}

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];

        int prefProd = 1;
        for (int i = 0; i < n; i++) {
            ans[i] = prefProd;
            prefProd *= nums[i];
        }

        int suffProd = 1;
        for (int i = n - 1; i >= 0; i--) {
            ans[i] *= suffProd;
            suffProd *= nums[i];
        }

        return ans;

    }
}